### PR TITLE
GitHub-Style Footnotes

### LINKED ISSUE #1285: Markdown footnotes enhancement
# Enhancement

## Environment

Zola version: 0.12.2 compiled from git

## Have footnotes link back to where they were footnoted

It would be great if there was a way to have a link-back to the location of the footnote. So if the site visitor is reading an article, they can click the footnote and then click the little return icon to go back to where the footnote was declared.

### PR DESCRIPTION
**Pull request by DerDrodt**
This PR adds the option of rendering footnotes at the bottoms in GitHub-Style, based on the implementation of [notriddle](https://github.com/raphlinus/pulldown-cmark/blob/ad19879e6b4f411dbdd78ffa6a9cf429e35afe47/examples/footnote-rewrite.rs).

Closes #1285.

For instance, the following markdown

```md
### Footnotes

Footnote 1 link[^first].

Footnote 2 link[^second].

Duplicated footnote reference[^second].

[^first]: Footnote **can have markup**.

[^second]: Footnote text.
```
Gets rendered to:
```html
<h3 id="footnotes">Footnotes</h3>
<p>Footnote 1 link<sup class="footnote-reference" id="fr-first-1"><a href="#fn-first">[1]</a></sup>.</p>
<p>Footnote 2 link<sup class="footnote-reference" id="fr-second-1"><a href="#fn-second">[2]</a></sup>.</p>
<p>Duplicated footnote reference<sup class="footnote-reference" id="fr-second-2"><a href="#fn-second">[2]</a></sup>.</p>
<hr><ol class="footnotes-list"><li id="fn-first" class="footnote-definition">
<p>Footnote <strong>can have markup</strong>. <a href="#fr-first-1">↩</a></p>
</li>
<li id="fn-second" class="footnote-definition">
<p>Footnote text. <a href="#fr-second-1">↩</a> <a href="#fr-second-2">↩2</a></p>
</li>
</ol>
```

This style is disabled by default and can be turned off via the `markdown.bottom_footnotes` option in the `config.toml`.

### CODE REVIEW

**Comment by Keats** [comment]
Can you clean up the commit history?

**Comment by DerDrodt** [comment]
> Can you clean up the commit history?

Took some work but it's done. To make this easier for other contributors, you could merge the master branch into next.

File: components/markdown/src/markdown.rs [hunk 1]
```diff
@@ -229,6 +231,156 @@ fn get_heading_refs(events: &[Event]) -> Vec<HeadingRef> {
...
+                    let html = Event::Html(format!(r##"<sup class="footnote-reference" id="fr-{name}-{nr}"><a href="#fn-{name}">[{n}]</a></sup>"##).into());
+                    if in_footnote.is_empty() {
+                        Some(html)
+                    } else {
+                        in_footnote.last_mut().unwrap().push(html);
+                        None
+                    }
+                }
+                _ if !in_footnote.is_empty() => {
+                    in_footnote.last_mut().unwrap().push(event);None
```
Comment (Keats) [inline components/markdown/src/markdown.rs]: can you put the None in a new line?

File: components/markdown/src/markdown.rs [hunk 2]
```diff
@@ -229,6 +231,156 @@ fn get_heading_refs(events: &[Event]) -> Vec<HeadingRef> {
...
+                }
+                Event::End(Tag::FootnoteDefinition(_)) => {
+                    let mut f = in_footnote.pop().unwrap();
+                    f.push(event);
+                    footnotes.push(f);
+                    None
+                }
+                Event::FootnoteReference(name) => {
+                    let n = footnote_numbers.len() + 1;
+                    let (n, nr) = footnote_numbers.entry(name.clone()).or_insert((n, 0usize));
```
Comment (Keats) [inline components/markdown/src/markdown.rs]: what does nr stands for?

File: components/markdown/src/markdown.rs [hunk 3]
```diff
@@ -229,6 +231,156 @@ fn get_heading_refs(events: &[Event]) -> Vec<HeadingRef> {
     heading_refs
 }
 
+/// Extracts the footnotes as bottom-notes, in the style of GitHub.
+///
+/// Adapted from [notriddle](https://github.com/raphlinus/pulldown-cmark/blob/ad19879e6b4f411dbdd78ffa6a9cf429e35afe47/examples/footnote-rewrite.rs).
+fn render_bottom_footnotes(events: Vec<Event<'_>>) -> Vec<Event<'_>> {
```
Comment (Keats) [inline components/markdown/src/markdown.rs]: Can that be moved to another file?

File: components/markdown/src/markdown.rs (hunk 2 above)
Comment (Drodt) [reply components/markdown/src/markdown.rs]: `n` is the index of the footnote. `nr` is the number of references to this footnote.

File: components/markdown/src/markdown.rs (hunk 3 above)
Comment (Drodt) [reply components/markdown/src/markdown.rs]: Done.

File: components/markdown/src/markdown.rs (hunk 1 above)
Comment (Drodt) [reply components/markdown/src/markdown.rs]: Done.

**Comment by vbrandl** [comment]
I don't know, if this is out of scope for this change, but currently footnotes are generated like this:

```html
<div class="footnote-definition" id="bar"><sup class="footnote-definition-label">1</sup>
  <p>baz</p>
</div>
<div class="footnote-definition" id="baz"><sup class="footnote-definition-label">2</sup>
  <p>lul</p>
</div>
```

I would like to render successive footnotes in a box, which is currently not possible. Is it possible to change the generated HTML to wrap successive footnotes inside a `div` with a CSS class to allow styling around a list of footnotes?

```html
<div class="footnotes-list">
  <div class="footnote-definition" id="bar"><sup class="footnote-definition-label">1</sup>
    <p>baz</p>
  </div>
  <div class="footnote-definition" id="baz"><sup class="footnote-definition-label">2</sup>
    <p>lul</p>
  </div>
</div>
```

That way, the `footnotes-list` class can be used to draw a box around any list of footnotes.

**Comment by Drodt** [comment]
> That way, the `footnotes-list` class can be used to draw a box around any list of footnotes.

I'm not a 100% sure, I understand you correctly, but all footnotes in this new style are wrapped in an `<ol>`. You can set a border on this and have your box.

**Comment by vbrandl** [comment]
I'm asking, if the *old* style of generating footnotes could be changed to wrap the footnotes in a `div` with class `footnotes-list`

**Comment by Drodt** [comment]
> I'm asking, if the _old_ style of generating footnotes could be changed to wrap the footnotes in a `div` with class `footnotes-list`

For that you should open a new Issue imo.

**Comment by vbrandl** [comment]
Ok, so it's out of scope, as expexted. I'll look into this.

Another question: shouldn't the Github style footnotes use the same CSS class names as the old footnotes?

**Comment by Keats** [comment]
> Another question: shouldn't the Github style footnotes use the same CSS class names as the old footnotes?

Probably? If it makes sense

**Comment by Drodt** [comment]
> Another question: shouldn't the Github style footnotes use the same CSS class names as the old footnotes?

I could do that, of course, but I don't think that's the best approach. As far as I can tell, quite a few markdown tools use the GitHub style with these classes. In my opinion, we should follow this more standard approach.

**Comment by Keats** [comment]
> quite a few markdown tools use the GitHub style with these classes

do you have examples?

**Comment by DerDrodt** [comment]
Ah, sorry. I was thinking about a different thing and mixed up the classes you were referring to. I have now added the `footnote-definition` class.

**Comment by DerDrodt** [comment]
I also noticed that pulldown-cmark has added support for GitHub-style footnotes (raphlinus/pulldown-cmark#654). So the code changed by this feature can be removed if and when pulldown-cmark publishes a new release.

(The feature was added in June to the repository, the last version was published in May. It may take a while.)

**Comment by Keats** [comment]
Oh let's wait for that then if you don't mind, we would switch to it at the next release of pulldown-cmark anyway

**Comment by aterenin** [comment]
Just a note that it looks like the upstream PR was merged.

I would be very interested in having generated footnotes that use `ol` instead of `div` elements.

**Comment by nikhilweee** [comment]
Curious to know if this would be merged soon?

**Comment by Keats** [comment]
No, we'll update the library to add support when it's released

**Comment by dkales** [comment]
FYI, it seems that there are now new releases available: https://github.com/raphlinus/pulldown-cmark/releases/tag/v0.9.6.

**Comment by si14** [comment]
Unfortunately those releases don't include https://github.com/raphlinus/pulldown-cmark/pull/654 yet.

**Comment by kytta** [comment]
https://github.com/raphlinus/pulldown-cmark/releases/tag/v0.10.0 is now out and it has the relevant footnote issue included

**Comment by Keats** [comment]
Sweet! That's a long list of changes, anyone interested in doing a PR? I think there are quite a few changes to make to the current parser

**Comment by timonvo** [comment]
@Keats I've just sent https://github.com/getzola/zola/pull/2432 to update to the new pulldown_cmark v0.10.0 version, since I had already had most of the necessary changes ready in my personal fork.

**Comment by WaffleLapkin** [comment]
With https://github.com/getzola/zola/pull/2432 merged, can this move forward?

**Comment by Keats** [comment]
It should probably on its own PR, with an option in the config.toml that will enable the pulldown-cmark flag. Closing this PR as it's now outdated

**Comment by WaffleLapkin** [comment]
Is someone in the thread willing to make a new version of this PR?

**Comment by totikom** [comment]
I've started working on it, you can assign me to #1285.

**Pull request closed without merging**